### PR TITLE
[RAPPS][SHLWAPI][SDK] Un-escape URL filename

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Choosing the better filename.

JIRA issue: [CORE-19490](https://jira.reactos.org/browse/CORE-19490)

image

## Proposed changes

- Add `UrlUnescapeAndMakeFileNameValid` helper function.
- Use `UrlUnescapeW` and `PathIsValidCharW`.
- Add `PATH_CHAR_CLASS_...` flags for `PathIsValidCharA/W`.

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by HBelusca** [comment]
> UrlUnescapeAndFileNameCensorship 

Not sure "Censorship" is of any useful help here....

File: base/applications/rapps/loaddlg.cpp [hunk 1]
```diff
@@ -710,8 +745,12 @@ CDownloadManager::ThreadFunc(LPVOID param)
                 Path += APPLICATION_DATABASE_NAME;
                 break;
             case DLTYPE_APPLICATION:
-                Path += (LPWSTR)(p + 1); // use the filename retrieved from URL
+            {
+                CStringW str = p + 1; // use the filename retrieved from URL
```
Comment (HBelusca) [inline base/applications/rapps/loaddlg.cpp:749]: Just to be sure, is the CStringW constructor here making a _**copy**_ (memory allocation and characters copy) of the string pointed by `p + 1` ? Or is the str internally still pointing to the original buffer pointed by  `p + 1`  ?

File: base/applications/rapps/loaddlg.cpp [hunk 2]
```diff
@@ -80,6 +81,40 @@ LoadStatusString(DownloadStatus StatusParam)
...
+    for (auto& ch : szPath)
+    {
+        if (ch == 0)
+            break;
+
+        if (!PathIsValidCharW(ch, FILENAME_VALID_CHAR))
+            ch = L'_';
+    }
+
+    str = szPath;
```
Comment (HBelusca) [inline base/applications/rapps/loaddlg.cpp:115]: What's happening there? Is `str` now getting a full copy of the contents of the `szPath` buffer?

File: base/applications/rapps/loaddlg.cpp (hunk 1 above)
Comment (katahiromz) [reply base/applications/rapps/loaddlg.cpp:749]: This `p` is pointing to the last slash in a URL.
`p + 1` will be the next character of the last slash.

File: base/applications/rapps/loaddlg.cpp (hunk 2 above)
Comment (katahiromz) [reply base/applications/rapps/loaddlg.cpp:115]: Yes. `str` will get the copy of `szPath`.

Comment (HBelusca) [reply base/applications/rapps/loaddlg.cpp:115]: ok thanks for the clarification

File: base/applications/rapps/loaddlg.cpp [hunk 3]
```diff
@@ -80,6 +81,40 @@ LoadStatusString(DownloadStatus StatusParam)
...
+    PATH_CHAR_CLASS_OTHER_VALID)
+
+VOID
+UrlUnescapeAndFileNameCensorship(CStringW& str)
+{
+    WCHAR szPath[MAX_PATH];
+    DWORD cchPath = _countof(szPath);
+    UrlUnescapeW(const_cast<LPWSTR>((LPCWSTR)str), szPath, &cchPath, 0);
+
+    for (auto& ch : szPath)
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply base/applications/rapps/loaddlg.cpp]: Fixed.

File: base/applications/rapps/loaddlg.cpp [hunk 4]
```diff
@@ -80,6 +81,40 @@ LoadStatusString(DownloadStatus StatusParam)
     return szString;
 }
 
+#define PATH_CHAR_CLASS_LETTER      0x00000001
+#define PATH_CHAR_CLASS_DOT         0x00000004
+#define PATH_CHAR_CLASS_SEMICOLON   0x00000020
+#define PATH_CHAR_CLASS_COMMA       0x00000040
+#define PATH_CHAR_CLASS_SPACE       0x00000080
+#define PATH_CHAR_CLASS_OTHER_VALID 0x00000100
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply base/applications/rapps/loaddlg.cpp]: Fixed.

File: base/applications/rapps/loaddlg.cpp [hunk 5]
```diff
@@ -80,6 +81,30 @@ LoadStatusString(DownloadStatus StatusParam)
...
+VOID
+UrlUnescapeAndMakeFileNameValid(CStringW& str)
+{
+    WCHAR szPath[MAX_PATH];
+    DWORD cchPath = _countof(szPath);
+    UrlUnescapeW(const_cast<LPWSTR>((LPCWSTR)str), szPath, &cchPath, 0);
+
+    for (PWCHAR pch = szPath; *pch; ++pch)
+    {
+        if (PathIsValidCharW(*pch, FILENAME_VALID_CHAR))
```
Comment (HBelusca) [inline base/applications/rapps/loaddlg.cpp:101]: Shouldn't this be
```suggestion
        if (!PathIsValidCharW(*pch, FILENAME_VALID_CHAR))
```
if this loop is meant to imitate what's done here:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/shell32/wine/shellpath.c;hb=dc4992ce4addc594722b5d23e1c137429b995b7d#l224
??

Comment (katahiromz) [reply base/applications/rapps/loaddlg.cpp]: Thanks.

**Pull request merged**